pkg/controller/nodelink/nodelink_controller_test.go

```diff
@@ -0,0 +1,649 @@
+		description             string
+		nodeTaints              []corev1.Taint
+		machineTaints           []corev1.Taint
+		expectedFinalNodeTaints []corev1.Taint
+	}{
+		{
+			description:             "no previous taint on node. Machine adds none",
+			nodeTaints:              []corev1.Taint{},
+			machineTaints:           []corev1.Taint{},
+			expectedFinalNodeTaints: []corev1.Taint{},
+		},
+		{
+			description:             "no previous taint on node. Machine adds one",
+			nodeTaints:              []corev1.Taint{},
+			machineTaints:           []corev1.Taint{{Key: "dedicated", Value: "some-value", Effect: "NoSchedule"}},
+			expectedFinalNodeTaints: []corev1.Taint{{Key: "dedicated", Value: "some-value", Effect: "NoSchedule"}},
+		},
+		{
+			description:   "already taint on node. Machine adds another",
+			nodeTaints:    []corev1.Taint{{Key: "key1", Value: "some-value", Effect: "Schedule"}},
+			machineTaints: []corev1.Taint{{Key: "dedicated", Value: "some-value", Effect: "NoSchedule"}},
+			expectedFinalNodeTaints: []corev1.Taint{{Key: "key1", Value: "some-value", Effect: "Schedule"},
+				{Key: "dedicated", Value: "some-value", Effect: "NoSchedule"}},
+		},
+		{
+			description:             "already taint on node. Machine adding same taint",
+			nodeTaints:              []corev1.Taint{{Key: "key1", Value: "v1", Effect: "Schedule"}},
+			machineTaints:           []corev1.Taint{{Key: "key1", Value: "v2", Effect: "Schedule"}},
+			expectedFinalNodeTaints: []corev1.Taint{{Key: "key1", Value: "v1", Effect: "Schedule"}},
+		},
+	}
+
+	for _, test := range testCases {
+		machine := machine("", "", nil, test.machineTaints)
+		node := node("", "", nil, test.nodeTaints)
+		addTaintsToNode(node, machine)
+		if !reflect.DeepEqual(node.Spec.Taints, test.expectedFinalNodeTaints) {
+			t.Errorf("Test case: %s. Expected: %v, got: %v", test.description, test.expectedFinalNodeTaints, node.Spec.Taints)
+		}
+	}
+}
+
+func TestNodeRequestFromMachine(t *testing.T) {
+	testCases := []struct {
+		machine  *mapiv1beta1.Machine
+		node     *corev1.Node
+		expected []reconcile.Request
+	}{
+		{
+			machine:  machine("noMatch", "", nil, nil),
+			node:     node("noMatch", "", nil, nil),
+			expected: []reconcile.Request{},
+		},
+		{
+			machine: machine("matchProviderID", "match", nil, nil),
+			node:    node("matchProviderID", "match", nil, nil),
+			expected: []reconcile.Request{
+				{
+					NamespacedName: client.ObjectKey{
+						Namespace: metav1.NamespaceNone,
+						Name:      "matchProviderID",
+					},
+				},
+			},
+		},
+		{
+			machine: machine("matchInternalIP", "", []corev1.NodeAddress{{
+				Type:    corev1.NodeInternalIP,
+				Address: "matchingInternalIP",
+			}}, nil),
+			node: node("matchInternalIP", "", []corev1.NodeAddress{{
+				Type:    corev1.NodeInternalIP,
+				Address: "matchingInternalIP",
+			}}, nil),
+			expected: []reconcile.Request{
+				{
+					NamespacedName: client.ObjectKey{
+						Namespace: metav1.NamespaceNone,
```

Comment:
test case with both, providerID and IP, matching?
test case with both, providerID and IP, non-empty and neither matching?